File: packages/sdk-python/src/forgemsg/webhook.py
"""
Webhook signature verification helper.

Usage::

    from forgemsg import verify_webhook_signature

    ok = verify_webhook_signature(
        secret=os.environ["FORGEMSG_WEBHOOK_SECRET"],
        payload=request.get_data(as_text=True),
        signature=request.headers.get("X-ForgeMsg-Signature", ""),
    )
    if not ok:
        abort(401)
"""

import hashlib
import hmac


def verify_webhook_signature(secret: str, payload: str, signature: str) -> bool:
    """
    Returns True if *signature* matches HMAC-SHA256(*secret*, *payload*).

    Uses :func:`hmac.compare_digest` for constant-time comparison.
    """
    expected = "sha256=" + hmac.new(
        secret.encode(),
        payload.encode(),
        hashlib.sha256,
    ).hexdigest()
    return hmac.compare_digest(expected, signature)

File: packages/sdk/src/webhook.ts
/**
 * Webhook signature verification helper.
 *
 * Usage:
 *   import { verifyWebhookSignature } from '@forgemsg/sdk';
 *
 *   app.post('/webhook', (req, res) => {
 *     const ok = verifyWebhookSignature({
 *       secret: process.env.FORGEMSG_WEBHOOK_SECRET,
 *       payload: req.rawBody,
 *       signature: req.headers['x-forgemsg-signature'],
 *     });
 *     if (!ok) return res.status(401).send('Invalid signature');
 *     // handle event ...
 *   });
 */

import { createHmac } from 'node:crypto';

export interface VerifyOptions {
  secret: string;
  /** Raw (unparsed) request body string */
  payload: string;
  /** Value of the X-ForgeMsg-Signature header */
  signature: string;
}

/**
 * Returns true if the signature matches HMAC-SHA256(secret, payload).
 * Uses constant-time comparison to prevent timing attacks.
 */
export function verifyWebhookSignature(opts: VerifyOptions): boolean {
  const expected = 'sha256=' + createHmac('sha256', opts.secret).update(opts.payload).digest('hex');

  if (expected.length !== opts.signature.length) return false;

  const a = Buffer.from(expected);
  const b = Buffer.from(opts.signature);
  // Constant-time XOR
  let diff = 0;
  for (let i = 0; i < a.length; i++) diff |= a[i] ^ b[i];
  return diff === 0;
}
